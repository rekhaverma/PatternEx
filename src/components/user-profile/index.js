export { default } from './user-profile.component.jsx';
