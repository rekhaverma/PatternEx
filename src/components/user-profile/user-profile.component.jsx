import React, { PureComponent } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { routerActions } from 'react-router-redux';

import { noAncestry } from 'lib';
import { restartZeppelin } from 'model/actions';

import ProfileGroup from './components/profile-group.component';

import './user-profile.style.scss';

const items = {
  'protected': [
    {
      'id': 'usersDashboardLink',
      'label': 'Users',
      'icon': 'icon-group',
    },
    {
      'id': 'rolesDashboardLink',
      'label': 'Roles',
      'icon': 'icon-cogs',
    },
    {
      'id': 'resourcesDashboardLink',
      'label': 'Resources',
      'icon': 'icon-server',
      'url': '/resources',
    },
  ],
  'settings': [
    {
      'id': 'settings',
      'label': 'Settings',
      'icon': '',
    },
  ],
  'user': [
    {
      'id': 'changePwd',
      'label': 'Change Password',
      'icon': '',
    },
    {
      'id': 'logOut',
      'label': 'Log Out',
      'icon': '',
    },
  ],
};

const models = {
  'protected': [
    // {
    //   'id': 'modelsDashboardLink',
    //   'label': 'Models',
    //   'icon': 'icon-code',
    // },
    {
      'id': 'rulesDashboardLink',
      'label': 'Rules',
      'icon': 'icon-list-ul',
      'url': '/rules',
    },
    {
      'id': 'tagsDashboardLink',
      'label': 'Tactics',
      'icon': 'icon-tags',
    },
    {
      'id': 'sysDumpLink',
      'label': 'System Dump',
      'icon': 'icon-download',
    },
    {
      'id': 'restartZeppelin',
      'label': 'Restart Zeppelin',
      'icon': 'icon-restart',
    },
    {
      'id': 'logSource',
      'label': 'Log Source',
      'icon': 'icon-download',
    },
  ],
};

export class UserProfile extends PureComponent {
  constructor(...args) {
    super(...args);

    this.state = {
      'boxIsOpen': false,
    };

    this.root = null;
    this.autoclose = this.autoclose.bind(this);
    this.closeBox = this.closeBox.bind(this);
    this.openBox = this.openBox.bind(this);
    this.selectOption = this.selectOption.bind(this);
  }

  componentWillUnmount() {
    this.detachEvent();
  }

  selectOption(option) {
    const redirects = [
      'usersDashboardLink',
      'rolesDashboardLink',
      'tagsDashboardLink',
    ];

    if (option === 'modelsDashboardLink') {
      window.location.href = `${window.location.origin}/models`;
    }

    if (option === 'logOut') {
      this.props.updateLocation('/logout');
    }

    if (option === 'changePwd') {
      this.props.setPopup('changePwd');
    }

    if (option === 'sysDumpLink') {
      window.open(`${window.location.origin}/sysdump`, '_blank');
    }

    if (option === 'settings') {
      this.props.setPopup('settings');
    }

    if (option === 'logSource') {
      this.props.setPopup('logSource');
    }

    if (option === 'restartZeppelin') {
      this.props.restartZeppelin();
    }

    if (redirects.includes(option)) {
      localStorage.setItem('route', option);
      window.location.href = window.location.origin;
    }
  }

  openBox() {
    this.setState({
      'boxIsOpen': true,
    }, () => {
      this.attachEvent();
    });
  }

  closeBox() {
    this.setState({
      'boxIsOpen': false,
    }, () => {
      this.detachEvent();
    });
  }

  autoclose(ev) {
    if (noAncestry(ev.target, ReactDOM.findDOMNode(this.root))) {
      this.closeBox();
    }
  }

  attachEvent() {
    document.addEventListener('click', this.autoclose);
  }

  detachEvent() {
    document.removeEventListener('click', this.autoclose);
  }

  render() {
    const data = this.props.type === 'user' ? items : models;
    return (
      <div className={this.props.className}>
        <div className="userProfile__iconGroup" onClick={this.openBox}>
          <span className={this.props.icon} />
          <span className="icon-chevron-down" />
        </div>
        <div
          className="profileBox"
          ref={ref => this.root = ref}
          style={{ 'display': this.state.boxIsOpen ? 'block' : 'none' }}
        >
          {
            Object.keys(data).map(key => (
              <ProfileGroup
                key={key}
                items={data[key]}
                privileges={this.props.privileges}
                type={key}
                onClick={this.selectOption}
                onClose={this.closeBox}
              />
            ))
          }
        </div>
      </div>
    );
  }
}
UserProfile.displayName = ' UserProfile';
UserProfile.propTypes = {
  'icon': PropTypes.string,
  'privileges': PropTypes.object.isRequired,
  'type': PropTypes.oneOf([
    'user',
    'configuration',
  ]),
  'setPopup': PropTypes.func,
  'className': PropTypes.string,
  'restartZeppelin': PropTypes.func.isRequired,
  'updateLocation': PropTypes.func.isRequired,
};
UserProfile.defaultProps = {
  'icon': 'icon-profile',
  'type': 'user',
  'setPopup': () => null,
  'className': 'userProfile',
};

const mapDispatchToProps = dispatch => ({
  'restartZeppelin': () => dispatch(restartZeppelin()),
  'updateLocation': location => dispatch(routerActions.push(location)),
});

export default connect(null, mapDispatchToProps)(UserProfile);
