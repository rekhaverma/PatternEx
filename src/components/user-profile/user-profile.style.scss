@import 'xbem';
@import 'theme';

@include block('userProfile') {
  position: relative;

  @include prefix(user-select, none);

  .icon-gear {
    font-size: setFontSizeNew('h2', 'big');
  }

  .icon-chevron-down {
    margin-left: 10px;
    font-size: setFontSizeNew('h5');
    font-weight: 300;
  }

  @include element('iconGroup') {
    @include flex($direction: row);
    align-items: center;
    justify-content: space-between;

    padding: 2px 8px 2px 2px;
    border-radius: 26px;

    @include prefix(transition, background-color .4s linear);

    cursor: pointer;

    &:hover {
      background: #191919; // todo: use function to set the color
    }
  }
}

@include block('profileBox') {
  position: absolute;
  top: 55px;
  right: -10px;
  width: 160px;
  padding: 10px;
  border: 1px solid setColor('black', 0.15);
  border-radius: 4px;
  box-shadow: 0 0 10px 0 #222; // todo: use function to set the color
  background-color: #2E3742; // todo: use function to set the color

  &:before {
    content: "";
    width: 20px;
    height: 20px;
    border-bottom: 10px solid #1e2125; // todo: use function to set the color
    border-top: 10px solid transparent;
    border-left: 10px solid transparent;
    border-right: 10px solid transparent;
    position: absolute;
    top: -40px;
    right: 10px;
  }

  @include element('group') {
    @include flex($direction: column);

    &:not(:first-child) {
      &:before {
        content: "";
        display: block;
        margin-top: 5px;
        margin-bottom: 5px;
        border-top: 1px solid #8A88A7; // todo: use function to set the color
      }
    }
  }

  @include element('item') {
    @include prefix(transition, color 0.4s);
    padding: 5px 0;
    margin: 0;
    color: #337ab7; // todo: use function to set the color
    font-size: setFontSizeNew('h4');
    cursor: pointer;

    @include setFontFamily('firaSans');

    [class^="icon-"], [class*=" icon-"]  {
      padding-right: 5px;
    }

    &:hover {
      color: setColor('white', 0.6);
    }

    @include state('protected') {
      text-decoration: none;
      color: setColor('white', 0.6);
      &:hover {
        color: #337ab7; // todo: use function to set the color
      }
    }
  }
}
