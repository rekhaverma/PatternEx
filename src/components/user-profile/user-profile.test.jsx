import React from 'react';
import sinon from 'sinon';


import { UserProfile } from './user-profile.component';

describe('<UserProfile />', () => {
  let wrapper;
  let props;
  const restartZeppelin = sinon.spy();

  beforeEach(() => {
    props = {
      icon: 'testIcon',
      privileges: {},
      type: 'user',
      setPopup: jest.fn(),
      className: 'testClass',
      restartZeppelin,
      updateLocation: jest.fn(),
    };
    wrapper = shallow(<UserProfile {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
