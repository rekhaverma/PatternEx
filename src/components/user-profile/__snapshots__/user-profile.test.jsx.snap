// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfile /> should render correctly 1`] = `
<div
  className="testClass"
>
  <div
    className="userProfile__iconGroup"
    onClick={[Function]}
  >
    <span
      className="testIcon"
    />
    <span
      className="icon-chevron-down"
    />
  </div>
  <div
    className="profileBox"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <ProfileGroup
      items={
        Array [
          Object {
            "icon": "icon-group",
            "id": "usersDashboardLink",
            "label": "Users",
          },
          Object {
            "icon": "icon-cogs",
            "id": "rolesDashboardLink",
            "label": "Roles",
          },
          Object {
            "icon": "icon-server",
            "id": "resourcesDashboardLink",
            "label": "Resources",
            "url": "/resources",
          },
        ]
      }
      key="protected"
      onClick={[Function]}
      onClose={[Function]}
      privileges={Object {}}
      type="protected"
    />
    <ProfileGroup
      items={
        Array [
          Object {
            "icon": "",
            "id": "settings",
            "label": "Settings",
          },
        ]
      }
      key="settings"
      onClick={[Function]}
      onClose={[Function]}
      privileges={Object {}}
      type="settings"
    />
    <ProfileGroup
      items={
        Array [
          Object {
            "icon": "",
            "id": "changePwd",
            "label": "Change Password",
          },
          Object {
            "icon": "",
            "id": "logOut",
            "label": "Log Out",
          },
        ]
      }
      key="user"
      onClick={[Function]}
      onClose={[Function]}
      privileges={Object {}}
      type="user"
    />
  </div>
</div>
`;
